Source/Particles/Collision/BinaryCollision/BinaryCollision.H

```diff
@@ -447,25 +470,98 @@ public:
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> pz_vec;
+            if (m_correct_energy_momentum) {
+                ww_weighted_sum_vec.resize(n_cells, 0.0_prt);
+                KE_vec.resize(n_cells, 0.0_prt);
+                px_vec.resize(n_cells, 0.0_prt);
+                py_vec.resize(n_cells, 0.0_prt);
+                pz_vec.resize(n_cells, 0.0_prt);
+            }
+            amrex::ParticleReal* AMREX_RESTRICT ww_weighted_sum_in_each_cell = ww_weighted_sum_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT KE_in_each_cell = KE_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT px_in_each_cell = px_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT py_in_each_cell = py_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT pz_in_each_cell = pz_vec.dataPtr();
+
+            // Save the momentum before the collision since sometimes the correction fails
+            // and the only solution is to restore the pre-collision values.
+            // The implicit solver already has ux_n etc so use those if available,
+            // otherwise create temporaries.
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> u1x_before;
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> u1y_before;
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> u1z_before;
+            amrex::ParticleReal* AMREX_RESTRICT u1x_before_ptr = nullptr;
+            amrex::ParticleReal* AMREX_RESTRICT u1y_before_ptr = nullptr;
+            amrex::ParticleReal* AMREX_RESTRICT u1z_before_ptr = nullptr;
+            if (m_correct_energy_momentum) {
+                // Check if ux_n was added.
+                std::vector<std::string> const & real_names1 = species_1.GetRealSoANames();
+                auto const pos1 = std::find(real_names1.begin(), real_names1.end(), "ux_n");
+                if (pos1 != real_names1.end()) {
+                    int const u1x_ni = std::distance(real_names1.begin(), pos1);
+                    u1x_before_ptr = soa_1.m_runtime_rdata[u1x_ni];
+                    u1y_before_ptr = soa_1.m_runtime_rdata[u1x_ni + 1];
+                    u1z_before_ptr = soa_1.m_runtime_rdata[u1x_ni + 2];
+                } else {
+                    u1x_before.resize(np1);
+                    u1y_before.resize(np1);
+                    u1z_before.resize(np1);
+                    u1x_before_ptr = u1x_before.dataPtr();
+                    u1y_before_ptr = u1y_before.dataPtr();
+                    u1z_before_ptr = u1z_before.dataPtr();
+                }
+
+            }
+
+            amrex::ParticleReal const beta_weight_exponent = m_beta_weight_exponent;
+
             WARPX_PROFILE_VAR("BinaryCollision::doCollisionsWithinTile::findDensityTemperatures", prof_findDensityTemperatures);
-        WARPX_PROFILE_VAR("BinaryCollision::doCollisionsWithinTile::findDensityTemperatures::atomics", prof_findDensityTemperatures_atomics);
+
+            if (m_correct_energy_momentum) {
```

Comment:
I have a similar question as before now. Here there is a loop over particles inside `if (m_correct_energy_momentum)`, and then immediately afterwards there is another loop over the particles with two `if` blocks inside the kernel. Would it be more performant to merge these two kernels and just add a third `if` block to the existing kernel, or should the existing kernel rather be split into two each inside their respective `if` blocks? Let me know if the question is not clear.

Reply:
I understand the question, but don't know the answer. There's the tradeoff of extra kernel launches versus a more complex kernel. I can do some timings to see is there is a difference. @atmyers would you have any intuition?

Reply:
I did some tests on Lassen and the combined kernel was definitely faster, by around 40%, presumably because of the overhead of the kernel launching. I implemented the combined kernels. Note that I also added the `if` check around the `ParallelFor` checking if any of the flags are set.

Reply:
Great! That's good to know. Thanks for taking the time to do the tests.